source/comp/markv_codec.cpp

```diff
@@ -0,0 +1,1194 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <numeric>
+#include <string>
+#include <vector>
+
+#include "binary.h"
+#include "diagnostic.h"
+#include "enum_string_mapping.h"
+#include "extensions.h"
+#include "instruction.h"
+#include "opcode.h"
+#include "operand.h"
+#include "spirv-tools/libspirv.h"
+#include "spirv-tools/markv.h"
+#include "spirv_endian.h"
+#include "spirv_validator_options.h"
+#include "util/bit_stream.h"
+#include "util/parse_number.h"
+#include "validate.h"
+#include "val/instruction.h"
+#include "val/validation_state.h"
+
+using libspirv::Instruction;
+using libspirv::ValidationState_t;
+using spvtools::ValidateInstructionAndUpdateValidationState;
+using spvutils::BitReaderWord64;
+using spvutils::BitWriterWord64;
+
+namespace {
+
+const uint32_t kSpirvMagicNumber = 0x07230203;
+const uint32_t kMarkvMagicNumber = 0x07230303;
+
+const size_t kCommentNumWhitespaces = 2;
+
+// TODO: This is a placeholder for a protobuf containing MARK-V model for a
+// specific dataset.
+class MarkvModel {
+ public:
+  size_t opcode_chunk_length() const { return 8; }
+  size_t num_operands_chunk_length() const { return 3; }
+
+  size_t u16_chunk_length() const { return 4; }
+  size_t s16_chunk_length() const { return 4; }
+  size_t s16_block_exponent() const { return 6; }
+
+  size_t u32_chunk_length() const { return 8; }
+  size_t s32_chunk_length() const { return 8; }
+  size_t s32_block_exponent() const { return 10; }
+
+  size_t u64_chunk_length() const { return 8; }
+  size_t s64_chunk_length() const { return 8; }
+  size_t s64_block_exponent() const { return 10; }
+};
+
+const MarkvModel* GetDefaultModel() {
+  static MarkvModel model;
+  return &model;
+}
+
+// Returns chunk length used for variable length encoding of spirv operand
+// words. Returns zero if operand type corresponds to potentially multiple
+// words or a word which is not expected to profit from variable width encoding.
```

Comment:
> Zero means do not encode, so it's safe.

That should be in the comment.   Saying "... not expected to profit from variable width encoding" does not say quite enough.